kbmd/box.c

```diff
@@ -0,0 +1,1057 @@
+		*keyp =  NULL;
+		*keylenp =  0;
+	}
+
+	return (ERRF_OK);
+}
+
+static errf_t *
+add_hexkey(nvlist_t *nvl, const char *name, const uint8_t *key, size_t keylen)
+{
+	errf_t *ret = ERRF_OK;
+	size_t keystrlen = keylen * 2 + 1;
+	char keystr[keystrlen];
+
+	bzero(keystr, keystrlen);
+	tohex(key, keylen, keystr, keystrlen);
+	ret = envlist_add_string(nvl, name, keystr);
+	explicit_bzero(keystr, keystrlen);
+	return (ret);
+}
+
+static errf_t *
+log_tpl(const struct ebox_tpl *rcfg, boolean_t stage)
+{
+	errf_t *ret = ERRF_OK;
+	uint8_t *hash = NULL;
+	size_t hashlen = 0;
+
+	if ((ret = template_hash(rcfg, &hash, &hashlen)) != ERRF_OK) {
+		(void) bunyan_error(tlog, "Failed to hash recovery config",
+		    BUNYAN_T_END);
+		return (ret);
+	}
+
+	char hashstr[hashlen * 2 + 1];
+
+	tohex(hash, hashlen, hashstr, hashlen * 2 + 1);
+
+	(void) bunyan_info(tlog, "Adding recovery configuration",
+	    BUNYAN_T_STRING, "staged", stage ? "true" : "false",
+	    BUNYAN_T_STRING, "hash", hashstr,
+	    BUNYAN_T_END);
+
+	free(hash);
+
+	return (ret);
+}
+
+errf_t *
+add_recovery(const char *dataset, const struct ebox_tpl *rcfg, boolean_t stage,
+    const recovery_token_t *rtoken)
+{
+	errf_t *ret = ERRF_OK;
+	kbmd_token_t *kt = NULL;
+	struct ebox *ebox = NULL;
+	const char *propstr = stage ? STAGEBOX_PROP : BOX_PROP;
+	char *eboxstr = NULL;
+	nvlist_t *zcp_args = NULL;
+	nvlist_t *result = NULL;
+	uint8_t *key = NULL;
+	size_t keylen = 0;
+
+	if (rcfg == NULL) {
+		return (errf("ArgumentError", NULL,
+		    "no recovery configuration given"));
+	}
+
+	if (rtoken != NULL && !RECOVERY_TOKEN_INRANGE(rtoken)) {
+		return (errf("RangeError", NULL,
+		    "incorrect recovery token size (%zu) out of range; "
+		    "must be at least in range [%u, %u] bytes",
+		    rtoken->rt_len,
+		    RECOVERY_TOKEN_MINLEN, RECOVERY_TOKEN_MAXLEN));
+	}
+
+	if ((ret = log_tpl(rcfg, stage)) != ERRF_OK) {
+		return (ret);
+	}
+
+	mutex_enter(&piv_lock);
+
+	if (sys_piv == NULL) {
+		mutex_exit(&piv_lock);
+		return (errf("UnlockError", NULL,
+		    "system zpool dataset must be set and unlocked before "
+		    "updating its recovery template"));
+	}
+	kt = sys_piv;
+
+	if (rtoken != NULL) {
+		if ((ret = set_piv_rtoken(kt, rtoken)) != ERRF_OK) {
+			goto done;
+		}
+	} else {
+		if ((ret = new_recovery_token(kt)) != ERRF_OK)
+			goto done;
+	}
+
+	if ((ret = kbmd_create_ebox(kt, rcfg, dataset, &key, &keylen,
+	    &ebox)) != ERRF_OK ||
+	    (ret = ebox_to_str(ebox, &eboxstr)) != ERRF_OK)
+		goto done;
+
+	if ((ret = envlist_alloc(&zcp_args)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "dataset",
+	    dataset)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "prop", propstr)) != ERRF_OK ||
+	    (ret = envlist_add_string(zcp_args, "ebox", eboxstr)) != ERRF_OK)
+		goto done;
+
+	if (!stage) {
+		if ((ret = add_hexkey(zcp_args, "keyhex", key,
+		    keylen)) != ERRF_OK) {
+			goto done;
+		}
+	}
+
+	ret = run_channel_program(dataset, add_prog, zcp_args, &result);
```

Comment:
I'll add a comment -- a zcp is used (instead of just setting the property) so when we're directly replacing the active ebox (instead of staging first), we can also set the key at the same time (atomically).

Reply:
I see that add_prog is initialized as:

```
  31 extern const char add_prog_start;
  32
  33 static const char *add_prog = &add_prog_start;
```

I can find no definition of `add_prog_start`.  I suspect that this is somehow related to the following in `Makefile`:

```
239 $(ZCP_OBJS): $(ZCP_SRCS_NUL)
240         $(ELFWRAP) -64 -o $@ $(ZCP_SRCS_NUL)
```

A comment is needed somewhere that explains this magic.  Probably near line 31.

Reply:
Added comments

Reply:
confirmed